src/ring_theory/chain_of_divisors.lean

```diff
@@ -0,0 +1,272 @@
+/-
+Copyright (c) 2021 . All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Anne Baanen, Paul Lezeau
+-/
+
+import algebra.is_prime_pow
+import algebra.squarefree
+
+/-!
+
+# Chains of divisors
+
+The results in this file show that in the monoid `associates M` of a `unique_factorization_monoid`
+`M`, an element `a` is an n-th prime power iff its set of divisors is a strictly increasing chain
+of length `n + 1`, aka we can find a strictly increasing bijection bewteen `fin (n + 1)` and
+the set of factors of `a`.
+
+We can then use this to show that given `a b : M`, if there is an monotone bijection
+between the sets of factors of `associates.mk a` and `associates.mk b`, then the prime
+factorisations of `a` and `b` have the same shape.
+
+
+## Main results
+- `prime_pow_has_chain` : existence of chain for prime powers
+- `is_prime_pow_of_has_chain` : elements that have a chain are prime powers
+- `multiplicity_prime_le_multiplicity_image_by_factor_order_iso` : if there is a monotone bijection
+  `d` between the set of factors of `a : associates M` and the set of factors of `b : associates N`,
+  then, for any prime `p ∣ a`, `multiplicity p a ≤ multiplicity (d p) b`.
+
+## Todo
+- show that under the assumptions of `multiplicity_prime_le_multiplicity_image_by_factor_order_iso`,
+  `d p` is prime. Applying `multiplicity_prime_le_multiplicity_image_by_factor_order_iso` on
+  `d.symm` then gives us `multiplicity p a = multiplicity (d p) b`
+- Maybe create a structure for chains of divisors?
+-/
+
+
+variables {M : Type*} [cancel_comm_monoid_with_zero M]
+
+open unique_factorization_monoid multiplicity irreducible
+
+namespace divisor_chain
+
+lemma pow_prime_has_chain {p : associates M} (n : ℕ) (hn : n ≠ 0) (hp : prime p) :
+  ∃ c : fin (n + 1) → associates M,
+    c 1 = p ∧ strict_mono c ∧
+    ∀ {r : associates M}, r ≤ p^n ↔ ∃ i, r = c i :=
+begin
+  refine ⟨λ i, p^(i : ℕ), _, λ n m h, _, λ y, ⟨λ h, _, _⟩⟩,
+  { rw [fin.coe_one', nat.mod_eq_of_lt, pow_one],
+    exact nat.lt_succ_of_le (nat.one_le_iff_ne_zero.mpr hn) },
+  { exact associates.dvd_not_unit_iff_lt.mp ⟨pow_ne_zero n hp.ne_zero, p^(m - n : ℕ),
+      not_is_unit_of_not_is_unit_dvd hp.not_unit (dvd_pow dvd_rfl (nat.sub_pos_of_lt h).ne'),
+      (pow_mul_pow_sub p h.le).symm⟩ },
+  { obtain ⟨i, i_le, hi⟩ := (dvd_prime_pow hp n).1 h,
+    rw associated_iff_eq at hi,
+    exact ⟨⟨i, nat.lt_succ_of_le i_le⟩, hi⟩ },
+  { rintro ⟨i, rfl⟩,
+    exact ⟨p^(n - i : ℕ), (pow_mul_pow_sub p (nat.succ_le_succ_iff.mp i.2)).symm⟩ }
+end
+
+lemma upper_chain_not_is_unit (n : ℕ) (i : fin (n + 1)) (i_pos : i ≠ 0)
```

Comment:
```suggestion
lemma upper_chain_not_is_unit {n : ℕ} {i : fin (n + 1)} (i_pos : i ≠ 0)
```
Please make all the variables that can be inferred by other assumptions implicit (this happens several times before). And again, you should find a better name (I know, it's annoying...).